hw: i386: virt: Add support for PCI hotplug through ACPI

sboeuf opened the pull request:
This PR enables the support for hotplugging PCI devices through ACPI:
- Enable PCI bus to accept hotplugged devices, by providing an
appropriate hotplug handler.
- Implement the code path generating the appropriate interrupt in
case a new PCI device is getting added to the PCI bus.
- Modify the definition of the PCI bus in the DSDT table to make
sure a PCI scan of all the available slots can be performed when
the interrupts comes in.
- Modify GED device in the DSDT table, so that an interrupt related
to PCI will trigger a new PCI scan.

Part of this work include some refactoring and the removal of unneeded
code related to the pc machine type.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

rbradford committed:
build: Switch submodules to http://

Makes it easier to work under proxied environments.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
hw: i386: Add virtual hardware machine type

The i386 virt machine type is a minimal, legacy free
platform based on virtual hardware.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add CPU machine class handlers

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add machine state init handler

Empty for now.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add NMI handler

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Move all virt code into its own folder

To avoid multiple thousand lines of code files...

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: pc: Move memory map code in its own file

We will also move most generic memory setting code to memory.c.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
hw/char: Add SYS_BUS version of debug console

This removes the requirement to have an ISA bus. Turn on with:

-device sysbus-debugcon,iobase=0x402,chardev=fw -chardev stdio,id=fw

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw/i386: Allow Q35 machine type to use "sysbus-debugcon"

It is necesary to add sysbus-debugcon to the whitelist of devices that
the machine can use.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
i386: virt: Add initial memory API

For now, it only prepares and sets e820 up.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add KVM clock

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Factorize CPU routines

A few routines are now shared between pc_* and virt, including the CPU
init one.
We factorize those routines into an i386 specific file that is now used
by all x86 machines.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: acpi: Decouple the ACPI build from the PC machine type

Factorize all needed information into a machine type agnostic structure
in order to be able to build ACPI tables without depending on the pc or
q35 machine type.
This will allow us to reuse the ACPI build routines for the i386/virt
machine type.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Build initial ACPI tables

Based on the now machine type agnostic ACPI building APIs.
Those are initial ACPI tables only as we will refactor them out to
expose a HW reduced topology only.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Factorize fw_cfg initialisation code

The fw_cfg device initialisation code is moved into its own file and
made PC machine type agnostic.
We need to do this code factorization in order for all x86 machine types
to be able initialize and add a fw_cfg device, including the new virt
one.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add fw_cfg device

The fw_cfg device is a virtual ROM device for sharing machine type
information with the firmware/BIOS that QEMU loads.
The one that virt uses is currently minimal and not very dynamic,
but it should be enough to pass the right information down to
e.g. an OVMF firmware.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: fw: Add SMBIOS and MSR feature control to fw_cfg

We expand the fw_cfg device data with SMBIOS properties and a
description of the MSR feature control we want to expose to the
firmware.
OVMF will need this in order to properly initialize and boot a guest
Linux kernel.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

chao-p committed:
i386: acpi: build static _PRT

For lightweight machine types like virt or pc-lite with
short boot time requirement, it makes sense to build a static _PRT and
reduce the guest ACPI interpretation time for both the guest kernel and
firmware.

Signed-off-by: Chao Peng <chao.p.peng@linux.intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
acpi: expose data structurs and functions of BIOS linker loader

Expose some data structures and functions for BIOS linker and loader.
Direct kernel boot/no firmware machine types need to link and load ACPI
tables into the guest address space, and those functions and data
structures will help us implement that feature for the virt machine type
at first.

Signed-off-by: Haozhong Zhang <haozhong.zhang@intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: acpi: Link guest ACPI tables when there is no firmware

Traditionally, guest firmware is responsible for patching and
linking the ACPI tables generated by QEMU.
However, nofw (no firmware) machine types like e.g. virt can't rely on
a non existent firmware to do so and thus need to do that from QEMU
directly.

Signed-off-by: Haozhong Zhang <haozhong.zhang@intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

jtkukunas committed:
i386: kernel-loader: Add direct loading protected/long mode kernel APIs

Traditionally a PC follows the following booting steps:

QEMU -> BIOS -> bootloader -> kernel realmode code
    -> kernel protected/longmode code.

This process takes lots of time. For light virtualization platform we
let QEMU loads protected/long mode kernel directly, hence skipping
several steps and speeding up the whole booting. We do this by filling
the zero page per Linux booting protocol and then jumping to the kernel
protected/long mode entry. Registers and paging should also be put in
the correct state.

Signed-off-by: Jim Kukunas <james.t.kukunas@linux.intel.com>
Signed-off-by: Chao Peng <chao.p.peng@linux.intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: loader: Load ROM via callback function

A callback load_fn is added to struct Rom, which is used to load the
ROM. A default loader rom_default_loader is used in case that a ROM is
created with setting load_fn.

This is needed to implement a zero host copy kernel loading mechanism,
by using custom ROM loaders.

Signed-off-by: Haozhong Zhang <haozhong.zhang@intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: loader: Load guest ELF kernel with no host copy for virt platform

virt platform loads guest ELF kernel by reading its directly from file into
guest and does not maintain a copy in ELF ROM. This way, we can save
the memory for the host copy.

Signed-off-by: Haozhong Zhang <haozhong.zhang@intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Enable direct kernel boot

When passing 'nofw' to the machine type acceleration options, we will
not add a fw_cfg device and we will directly boot the provided kernel
image.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Allow sysbus console debug device

And add it to the default x86_64 config. This gives us
a basic serial output.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
pc: Refactor out PC CPU code

This CPU code does not rely on PCMachineState and so can be refactored
out for use with the virt machine.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
virt: Disable ACPI setup for now

This isn't necessary for booting direct boot and segfaults when CPUs are
correctly initialised.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
i386: virt: Implement CPU hotplug

This is necessary to get the CPU correctly initialised. Later on this
will need to be integrated with ACPI.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: i386: Make firmware loading non-PC specific

Minor refactoring and relocation of code to make the firmware loading
code not dependent on the PC specific code. This allows the code to be
used for the virt platform.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
i386: virt: Load firmware on virt platform

Load the firmware on the virt platform.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

chao-p committed:
pci: introduce light weight PCIE Host emulation pci-lite

A minimal emulation for PCIE host bridge that supports 0xcf8/0xcfc and
MMFG. Actually there is already a gpex there, but it is designed mainly
for ARM and is not quite suitable for x86:
- it lacks things like PCI hole properties which are required by ACPI.
- the corresponding driver in Linux is designed to work with device tree
  which is not suitable for x86. For this case, additional guest driver
  is even not needed.

Currently MMFG size is limited to 1M, which means only 1 bus is
supported, this is aimed to reduce the scan time in guest. And it
doesn't have a valid vendor ID/device ID assigned so guest may not
recognize it, the functionality however is expected to be OK.

Updated for newer QEMU by Rob Bradford.

Signed-off-by: Chao Peng <chao.p.peng@linux.intel.com>
Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
hw: acpi: Factor some i386 ACPI code out

We move some structure into the architecture agnostic
include/hw/acpi/acpi.h header, and we move the ACPI link
API into hw/acpi/link.c

We need to do so in preparation for the generic HW reduced ACPI
implementation.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
hw: pci-host: Fix pci-lite bridge for current QEMU

PCI devices are expected to inherit from an interface to indicate what type
that device is. Update pci-lite to indicate the host device is a conventional
PCI device.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: pci-host: Use GPEX details for pci-lite host device

Copy the PCI device details from the GPEX device into pci-lite. This might just
be a temporary solution until we allocate our own PCI ID.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: pci-host: Expose initialisation function in a header

Expose the device initialisation function in a header so that it can be used by
virt.c.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
i386: virt: Add PCI lite host bridge to machine

Add PCI lite host bridge to the virt machine platform and setup the IRQs for
it.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
i386: virt: Track low memory size through the VM state

We need this for the ACPI link to work. The ACPI link code checks that
there is enough space in lowmem for creating the ACPI zones.
If we don't track that field and leave it to 0, the check will always
fail.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Initial HW reduced API implementation

We build a minimal set of ACPI hw reduced tables: RSDT,
FADT, DSDT pointed by a RSDP.
The DSDT only contains one PCI host bridge.

This will be directly consumed by the i386 virt platform.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
tools: Add master branch CI back in

We need it to properly run our semaphore based CI.
This could partially be upstreamed as well.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: fw-build: Add firmware build methods and state

In order to factorize ACPI tables build routines across machine types
and also across architectures (ARM and i386), we create a new
FirmwareBuildMethods structure in order to abstract ACPI and DT methods.

Initially, this will allow us to customize ACPI tables building for different
machine types without duplicating a lot of ACPI build code.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
hw/i386: Add basic CMOS device for virt machine

This basic device provides a virtual CMOS I/O port mechanism for exposing
values to the guest. In particular OVMF expects to be able to read the basic
memory details from an emulated CMOS.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw/i386: Expose virt machine RAM details via CMOS

Expose the virtual platform basic RAM details via the virtual CMOS device.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
acpi: reduced: Pass ACPI configuration pointer to the build routine

So that we can then call the MADT build method with the ACPI
configuration data.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add virtual ACPI device

This is mostly empty for now, hanging off the system bus.
We need this in order to re-use the MADT i386 build routine for the
virt machine type, as it depends on a HotplugHandler pointer.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Initialize the virtual ACPI device

Our virt machine type is also linked to the virtual ACPI device.
This is needed for some of the ACPI build process, and also eventually
for doing CPU and memory hotplug, and ACPI based PM.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
acpi: reduced: Add MADT table

Now that our virt machine type correctrly points at a virtual ACPI
device, we can safely add the MADT/APIC table to the ACPI set.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Implement XSDT support for RSDP

XSDT is the 64-bit version of the legacy ACPI RSDT (Root System
Description Table). RSDT only allow for 32-bit addressses and have thus
been deprecated. Since ACPI version 2.0, RSDPs should point at XSDTs and
no longer RSDTs, although RSDTs are still supported for backward
compatibility.

Since version 2.0, RSDPs should add an extended checksum, a complete table
length and a version field to the table.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: arm: Switch to the AML build RSDP building routine

And fix the RSDP table. The previous implementation was missing adding
both the legacy and extended checksums.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
virt: pci_lite: change pci hole/hole64 related functions

Those changes can fix segfault in hole64 start/end functions.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
acpi: reduced: Pass the right object to the update callback

We are passing a AcpiBuildState object while the
acpi_reduced_build_update() routine takes a MachineState one as an
argument.
The joys of the C language.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
virt: acpi: get pcilite host bridge

Add the pcilite for acpi_get_i386_pci_host() and refactor this
function for general qemu usage.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
acpi: make aml build routines general

The virt/i386 acpi tables also use those routines and move
those into aml-build.c for general usage.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

rbradford committed:
hw: Use dynamic checking casts in callbacks

When casting from void * pointer in callbacks use the casts which
feature a dynamic type check to avoid issues with passing the wrong type
into data pointer for the callback.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
Revert "hw/i386: Expose virt machine RAM details via CMOS"

This reverts commit 0722f2c4218b5d64b25009465c39492a40321bbc.

rbradford committed:
Revert "hw/i386: Add basic CMOS device for virt machine"

This reverts commit c1533ffec3d8bf329926cb93002815bed6c05eaf.

Commit:
Revert "i386: acpi: build static _PRT"

This patch breaks q35 shutdown/reboot behaviour in nemu.
This reverts commit a201848e2c51a8b666cd8e63ddcf4bdc061eff40.

yangzhon committed:
virt: acpi: make build_osc_method() more general

The _OSC aml table is almost same in q35 and arm-virt,
So, build_osc_method() will be use in all pcie hostbridge
platform.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: acpi: add dsdt pci support

porting dsdt pci settings from pc/q35 into virt/i386.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
i386: virt: Hook the virt CPU hotplug into ACPI

With the virt machine type, only ACPI hotplug is supported for CPU and
memory hotplug.
This patch plugs the virt CPU hotplug handler into the ACPI framework,
through the virt ACPI device. It adds support for CPU unplug as well.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
virt: acpi: Implement ACPI CPU hotplug hooks

We register our ACPI device with the CPU hotplug framework and then
implement the needed hotplug/unplug device hooks. They only support CPU
hotplug for now.

Also, the virt DSDT table does not yet contain the CPU and GED AML so
the ACPI event sending won't trigger anything yet.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
virt: acpi: add mmcfg support

The pcie bus is default bus of virt/i386 platform, the mmcfg
need be supported in guest kernel.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

rbradford committed:
i386: virt: Use 2GiB for top of low ram

This is the same as Q35 and so makes the OVMF memory map setup
convenient.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: pci-lite: Put PCI Express BAR just above lowmem

Adjust the PCI Lite's configuration space so that it is the same size as
Q35 and also located above the lowmem limit.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
virt: add HW-reduced ACPI shutdown support

HW-reduced ACPI uses SLEEP_CONTROL_REG to enter S5 sleep state.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
acpi: reduced: GED AML generation routine

We generate the AML code for handling GEDs.
The build_ged_aml routine takes a GedEvent array that maps a specific
GED event to an IRQ number. Then we use that array to build both the
_CRS and the _EVT section of the GED device.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
hw/arm: Fix ARM build after elf loader change

The ARM build was failing as load_elf_as was modified to take an extra
parameter.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: acpi: Adjust function naming to avoid clashes

Some functions were named such that they clashed with other functions that were
part of the ARM build. Rename those functions to avoid the clashes.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: arm: Fix ACPI build

Remove unused variable after earlier refactoring.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
hw: acpi: Only include link.c in x86 builds

This is for x86 direct kernel boot only.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
tools: Fix ccloudvm ARM build to work in Semaphore

The version of host 9pfs in the Semaphore VM is old and causes problems when
trying to create files directly inside a mounted path. Instead copy the files
into the build.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
i386: virt: Decouple IOAPIC initialization from PCI

Our IOAPIC settings should not be tied to PCI.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
i386: virt: Add supported GED events

For now we only support CPU hotplug.
We add an array of events and use the first non legacy GSI IRQ for our
GED events. Given that we control the virt platform, we can safely use
those.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
tools: Download the latest OVMF binary from our fork

This binary is designed to support both Q35/PIIX and virt so can be used
in place of the Clear Linux one.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
tools/CI: Enable virt testing with Clear Linux

Start testing virt machine type with Clear Cloud.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
tools/CI: Exit non-zero if test running fails

If running a test fails, keep going, but exit with a non-zero code.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
virt: acpi: Implement the ACPI events handler

With virt we only support sending GED events, through dedicated IRQ
lines. Here we only support CPU hotplug by pulsing the corresponding
interrupt line.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
virt: add HW-reduced ACPI reboot support

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

yangzhon committed:
i386: virt: Hook the virt memory hotplug into ACPI

This patch add the virt memory hotplug/unplug handler into
the ACPI framework.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: acpi: Implement ACPI memory hotplug hooks

We register acpi device with memory hotplug framework and then
implement the hotplug/unplug device hooks.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
acpi: reduced: Separate GED device creation AML from the CPUs one

We are currently calling the GED AML build from the CPUs addition one.
They're not directlt related, so we should just separate them out.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
virt: acpi: add memory hotplug support in dsdt

We add the memory hotplug in dsdt acpi table and use GED
events to handle memory hotplug.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
i386: virt: Do not error out on hotplug pre-plug hook

Not all device types need a pre-plug hook implementation.
For example memory hotplug does not need one, so we should not error out
when getting a pre-plug request for such devices.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
acpi: Do not create hotplug method when handler is not defined

HW reduced ACPI handles all hotplug events through the GED device and
its methods do not need event specific methods.
So we make those event method names optional and create a corresponding
AML only when the method name is not NULL.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
virt: pci: change 64bit pci hole for memory hotplug

The previous 64bit pci hole used fixed value, the new implementation
add the memory hotplug.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
i386: virt: initialize hotplug memory region

Initialize hotplug memory address space for memory
hotplug usage.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

rbradford committed:
i386: virt: Initialise AcpiConfiguration pointer earlier

Previously the vms->acpi_configuration pointer was setup in the "done"
notification for machine initialisation. However this is too late when
memory configuration details are provided on the command line (and thus
the memory hotplug functions are used.)

As nothing in this struct relies on values that are set between
virt_machine_state_init and virt_machine_done then we can safely
initialise this this pointer and populate the structure during
virt_machine_state_init.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
hw: i386: virt: Add NVDIMM support

This commit introduces the NVDIMM support. It includes the NVDIMM
hotplug since the way NVDIMM is added is always through hotplug,
even when an NVDIMM device is defined from the Qemu command line.

First, this code builds a new SSDT table with the NVDR device
described inside as a specific platform device.

Then it updates the GED method of the DSDT table, by defining a new
Notify() method to notify the guest OS driver that something changed
regarding the NVDIMM devices.

All of this happen if NVDIMM is enabled at the machine level.

The last thing this patch does is to send an interrupt to the guest
kernel everytime a new NVDIMM device is plugged. This triggers the
GED method defined in the DSDT.

Fixes #44

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

yangzhon committed:
virt: acpi: fixed one cpu acpi bug

From the guest kernel log, we can find below error log:
acpi ACPI0007:08: BIOS reported wrong ACPI id 0 for the processor
acpi ACPI0007:09: BIOS reported wrong ACPI id 1 for the processor

This issue is caused by re-define ACPI0007 aml table. Those changes
do not impact cpu hotplug.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
virt: refactor the sleep state adding codes

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

yangzhon committed:
virt: acpi: move build_srat() to aml-build.c

Move build_srat() related function into aml-build.c for
different platform.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: acpi: add numa support

Build srat table to send the numa topology to the BIOS, which has to
setup the tables. The guest kernel can normally show numa acpi info.
Since the default numa code is still here, when we build srat table for
numa, we can adds a -numa command line parameter in virt platform and
expose those numa nodes to vm.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

rbradford committed:
i386: virt: Enforce DIMM memory restrictions

When hotplugging memory the Linux kernel will reject memory that is not
correctly aligned to 128MiB and expects the size to be a multiple of the
same. However for NVDIMM no such restrictions apply but in order to be
able to combine the hotplugging of NVDIMMs with memory hotplug it is
necessary to ensure that all NVDIMM start addresses are aligned as
above.

We achieve that by identifying the next available address for the memory
and rounding that to 128MiB and using that as the allocation hint.
Further if this is not an NVDIMM being plugged we also use the existing
alignment mechanism to enforce the size constraint too.

This was tested by adding a small NVDIMM and then hotplugging memory.
The appropriate output from /proc/iomem looks like this:

fee00000-fee00fff : Local APIC
100000000-10000bfff : Persistent Memory
  100000000-10000bfff : namespace0.0
108000000-147ffffff : System RAM

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

yangzhon committed:
acpi: memory-hotplug: fixed dsdt error in dsl file

The error log as below:
dsdt.dsl 266: Return (MOST (_UID, Arg0, Arg1, Arg2))
Error 6080 - Called method returns no value ^

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

sboeuf commented:
@liujing2 here is some work in progress on my side about PCI devices.
Just to summarize this, I can hotplug some pci devices to `pcie.0`, but I need to rescan manually the PCI bus from inside the guest to make the device appear.
I am working on the GED support, and how to make the ACPI trigger the rescan when it receives the interrupt.

liujing2 commented:
@sboeuf  wow, your codes almost done the key parts of acpi pci hotplug. I also prepared one #59 but have some differences and met problems now. And I prefer yours. 
I looked into the codes and got why we have to manually rescan. I wrote a patch for it and submitted PR #60. Now acpi pci hotplug onto pcie.0 can be done successfully. 
BTW, I think we need some function factoring.  I will take a look later today.

Commit:
hw: i386: virt: Add PCI hotplug through APCI

This first commit enables the major part of PCI hotplug through
APCI:
- It enables the PCI bus to accept hotplugged devices, by providing
  an appropriate hotplug handler.
- It implements the code path generating the appropriate interrupt
  in case a new PCI device is getting added to the PCI bus.
- It modifies the definition of the PCI bus in the DSDT table to
  make sure a PCI scan of all the available slots can be performed
  when the interrupts comes in.
- At last, it modifies GED device in the DSDT table, so that an
  interrupt related to PCI will trigger a new PCI scan.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
hw: acpi: Factorize ACPI definitions for PCI hotplug

This commit factorizes the code responsible of adding the PCI hotplug
ability to the PCI device in the DSDT table. This way, the code can
now be shared between virt and pc machine types.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
hw: acpi: pcihp: Retrieve the PCI bus from AcpiPciHpState

Instead of using the very non-generic method find_i440fx() to
retrieve the PCI bus, which works only for pc machine type, this
commits aims to rely on the fact that the PCI bus is known by the
structure AcpiPciHpState.

When the structure is initialized through acpi_pcihp_init() call,
it saves the PCI bus, which means there is no need to invoke a
special function later on.

Based on the fact that find_i440fx() was only used there, this
patch also removes the function find_i440fx() itself from the
entire codebase.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

sboeuf commented:
@liujing2 I have completed the rework of this PR, including some factorization and some cleanup. I think we can close #59 and #60 as this PR supersedes them. WDYT?

@sameo @mcastelino @rbradford @yangzhon @liujing2 PTAL
